src/Fantomas/CodePrinter.fs

```diff
@@ -1322,13 +1322,64 @@ and genExpr astContext synExpr =
 
-        atCurrentColumn
-            (genExpr astContext e
-             +> colPre sepSpace sepSpace es (fun e ->
-                    onlyIf (isCompExpr e) (sepSpace +> sepOpenSFixed +> sepSpace)
+        // This checks if there is an application with multiple parameters which would fit in one line
+        // but at least one of the parameters which is not the last parameter has a line comment 
+        // In that case we have to split the arguments over multiple lines to preserve the line comments
+        // Example:
+        //
+        //     someFunc a b c // comment
+        //
+        // This can stay as is because the last argument has the line comment
+        // However:
+        //
+        //     someFunc
+        //         a 
+        //         b // comment
+        //         c
+        //
+        // This has to be split over multiple lines to preserve the line comment
+        let parametersMustBeNewlinedBecauseOfTriviaContent es ctx =
+            es
+            |> List.map (fun (e:SynExpr) -> TriviaHelpers.findByRange ctx.Trivia e.Range)
+            |> List.mapi (fun i node -> if node.IsSome && not (i = List.length es - 1) then TriviaHelpers.``has line comment after`` node.Value else false)
+            |> List.contains true
+
+        fun ctx ->
+            let shortExpression = 
+                (atCurrentColumn
+                    (genExpr astContext e
+                    +> colPre sepSpace sepSpace es (fun e ->
+                    ifElse (isCompExpr e)
+                        (sepSpace
+                        +> sepOpenSFixed
+                        +> sepSpace
+                        +> indent
+                        +> appNlnFun e (indentIfNeeded +> genExpr astContext e)
+                        +> unindent)
+                        (indent
+                        +> indentIfNeeded
+                        +> sepNln
+                        +> genExpr astContext e
+                        +> unindent))))
+
+            let longExpression =
+                (atCurrentColumn
+                (genExpr astContext e
+                +> colPre sepSpace sepSpace es (fun e ->
+                ifElse (isCompExpr e)
+                    (sepSpace
+                    +> sepOpenSFixed
+                    +> sepSpace
                     +> indent
                     +> appNlnFun e (indentIfNeeded +> genExpr astContext e)
-                    +> unindent))
+                    +> unindent)
+                    (indent
+                    +> indentIfNeeded
+                    +> genExpr astContext e
+                    +> unindent))))
+
+            //ifElseCtx (fun ctx -> exceedsWidth ctx.Config.PageWidth longExpression ctx || parametersMustBeNewlinedBecauseOfTriviaContent es ctx) shortExpression longExpression ctx
+            ifElseCtx (fun ctx -> parametersMustBeNewlinedBecauseOfTriviaContent es ctx) shortExpression (expressionFitsOnRestOfLine longExpression shortExpression) ctx
```

Comment:
`expressionFitsOnRestOfLine longExpression shortExpression` that doesn't really make much sense.
The function `expressionFitsOnRestOfLine` will first try the longExpression and use the short one as fallback.
Those parameters should be swapped.

I also think you don't always want this behaviour, it will depend on the `es` of `| App(e, es)`.
If one expression in `es` is a lambda for example I don't think you necessarily want the suggested formatting of putting every argument on one line.

For example:
```fsharp
let foo =
    Array.map (fun x ->
                                  // some multiline code
                       x + x) [| 8;9;6;5 |]
```

I don't think you want to make 
```fsharp
let foo =
    Array.map
         (fun x ->
                       // some multiline code
           x + x)
     [| 8;9;6;5 |]
```
of it.